README.md
[![Build Status](https://travis-ci.org/wang-q/App-Fasops.svg?branch=master)](https://travis-ci.org/wang-q/App-Fasops) [![Coverage Status](http://codecov.io/github/wang-q/App-Fasops/coverage.svg?branch=master)](https://codecov.io/github/wang-q/App-Fasops?branch=master) [![MetaCPAN Release](https://badge.fury.io/pl/App-Fasops.svg)](https://metacpan.org/release/App-Fasops)
# NAME

App::Fasops - operating blocked fasta files

# SYNOPSIS

    fasops <command> [-?h] [long options...]
        -? -h --help    show help

    Available commands:

      commands: list the application's commands
          help: display a command's help screen

       axt2fas: convert axt to blocked fasta
         check: check genome locations in (blocked) fasta headers
        covers: scan blocked fasta files and output covers on chromosomes
          join: join multiple blocked fasta files by common target
         links: scan blocked fasta files and output bi/multi-lateral range links
       maf2fas: convert maf to blocked fasta
         names: scan blocked fasta files and output all species names
        refine: realign blocked fasta file with external programs
       replace: replace headers from a blocked fasta
      separate: separate blocked fasta files by species
         slice: extract alignment slices from a blocked fasta
         split: split blocked fasta files to separate per-alignment files
        subset: extract a subset of species from a blocked fasta
          xlsx: paint substitutions and indels to an excel file

See `fasops commands` for usage information.

# AUTHOR

Qiang Wang &lt;wang-q@outlook.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Qiang Wang.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for App-Fasops

{{$NEXT}}

0.6.4 2019-09-02T19:31:22Z

    - Move command `create` from `rangeops` to here
    - Move build_info() and get_seq_faidx() to Common.pm

0.6.3 2019-03-14T19:45:27Z

    - Add --min and --max to `fasops xlsx`

0.6.2 2019-02-01T23:35:08Z

    - Add parse_block_array() to Common.pm
    - `fasops refine`
        - Use array to avoid IxHash

0.6.1 2019-01-25T15:45:18Z

    - Add `fasops vars`

0.6.0 2018-04-22T14:40:50Z

    - Add `fasops consensus`
    - Add poa_consensus() to Common.pm
    - Require Perl 5.18 as previous versions failed to find shared file
    - Better description() for all commands

0.5.16 2018-04-16T16:07:23Z

    - Add --simple to `fasops split`

0.5.15 2018-04-01T17:27:20Z

    - Fix maf2fas for reverse-complemented strands

0.5.14 2018-03-28T04:55:06Z

    - Update help
    - Release to CPAN

0.5.13 2018-03-24T08:15:55Z

    - Add `fasops mergecsv`, moved from egaz

0.5.12 2017-09-26T11:54:31Z

    - Add --nosingle, --noindel and --nocomplex to `fasops xlsx`

0.5.11 2016-12-19T19:56:59Z

    - Add `fasops stat`

0.5.10 2016-12-18T20:19:59Z

    - Remove dependency on List::MoreUtils
    - Move calc_ld() here

0.5.9 2016-12-17T18:06:32Z

    - Close $out_fh and $in_fh at once

0.5.8 2016-12-17T15:06:56Z

    - Add `fasops concat`

0.5.7 2016-12-17T04:57:15Z

    - Add `fasops check`

0.5.6 2016-12-16T10:30:10Z

    - Fix unknown SNPs
    - Add snp_ prefix to get_snps() and polarize_snp()

0.5.5 2016-12-16T02:52:25Z

    - Switch to Minilla
    - Improve test coverage
    - Without outgroups, consider indels occurred at target

0.5.4  2016-12-15 03:13:17 CST

    - Add `fasops xlsx`
    - show_defaults => 1 for all commands

0.5.3  2016-12-14 05:41:39 CST

    - Add get_snps(), polarize_snp()
    - Add get_indels(), polarize_indel()
    - Merge xt/11-refine.t to t/

0.5.2  2016-12-12 02:06:09 CST

    - Bring Perl 5.10 back

0.5.1  2016-12-09 21:51:44 CST

    - Drop Perl 5.10

0.5.0  2016-12-09 21:22:25 CST

    - Read from stdin

0.4.13  2016-07-07 22:32:08 CST

    - Add codecov badge

0.4.12  2016-06-30 16:35:03 CST

    - Speed up Common::read_fasta()

0.4.11  2016-06-26 03:46:35 CST

    - Fix command refine with --chop

0.4.10  2016-06-26 03:20:10 CST

    - Add --chop to command refine

0.4.9  2016-06-25 22:03:35 CST

    - Merge command multilinks into links

0.4.8  2016-06-21 06:11:14 CST

    - Add --best to command links
    - Command multilinks

0.4.7  2016-06-19 04:06:59 CST

    - App::RL::Common::decode_header() changed

0.4.6  2016-05-30 21:31:20 CST

    - Improve command refine following guides provided by Mario Roy.

0.4.5  2016-05-24 22:33:02 CST

    - Better prompts

0.4.4  2016-05-24 08:36:40 CST

    - Clean prompts

0.4.3  2016-05-21 19:47:58 CST

    - Parallelize command refine

0.4.2  2016-05-14 03:54:19 CST

    - Add `fasops join`
    - Add --required to command subset
    - Add --chr to command split

0.4.1  2016-05-12 22:44:56 CST

    - Add --length to command slice
    - Add --length and --trim to command covers

0.4.0  2016-05-12 03:40:52 CST

    - Specify versions of dependencies

0.3.14  2016-05-12 02:15:06 CST

    - Add `fasops slice`

0.3.13  2016-05-10 21:00:06 CST

    - `fasops refine --outgroup --quick`

0.3.12  2016-05-09 17:14:03 CST

    - Fix ref_pair_D

0.3.11  2016-05-08 20:02:14 CST

    - Move subs from AlignDB::Util

0.3.10  2016-05-07 17:30:24 CST

    - codecov

0.3.9  2016-05-07 14:15:37 CST

    - Export nothing. Remove duplicated functions
    - Add strict and warnings to every modules

0.3.8  2016-05-05 00:26:00 CST

    - move 11-refine.t to xt/

0.3.7  2016-05-04 01:27:07 CST

    - Add `fasops refine`

0.3.6  2016-04-08 15:06:47 CST

    - Require 5.10

0.3.5  2016-04-08 15:01:07 CST

    - Add travis-ci

0.3.4  2016-02-27 01:57:33 CST

    - `fasops replace` remove blocks

0.3.3  2016-02-07 11:43:35 CST

    - Fix warnings

0.3.2  2016-01-15 17:18:28 CST

    - A real Changes file

0.3.1  Thu Jan  7 02:12:28 2016

    - Add --size for `fasops axt2fas`

0.2.9  Mon Dec 28 17:02:40 2015

    - Common functions
    - Add `fasops links`
    - Add `fasops replace`

0.2.6  Mon Dec 28 09:31:35 2015

    - Change regex in App::Fasops::decode_header()

0.2.5  Sun Dec 27 21:11:13 2015

    - Tests for `fasops separate`

0.2.4  Fri Dec 25 02:22:41 2015

    - Rewrite axt2fas and maf2fas

0.2.3  Thu Dec 24 23:24:10 2015

    - Add `fasops covers`
    - split and names accept more than one files

0.2.2  2015-12-24 22:09:28 CST

    - Fix `fasops split`
    - Add `fasops separate`

0.2.1  2015-10-28 19:19:10 CST

    - Add license

0.2.0  2015-07-31 02:49:59 CST

    - Convert to Dist::Milla

0.1.0  2015-04-13 00:00:00 CST

    - Now useful


lib/App/Fasops/Command/xlsx.pm
package App::Fasops::Command::xlsx;
use strict;
use warnings;
use autodie;

use Excel::Writer::XLSX;

use App::Fasops -command;
use App::Fasops::Common;

sub abstract {
    return 'paint substitutions and indels to an excel file';
}

sub opt_spec {
    return (
        [ "outfile|o=s", "Output filename. [stdout] for screen" ],
        [ "length|l=i", "the threshold of alignment length", { default => 1 } ],
        [ 'wrap=i',     'wrap length',                       { default => 50 }, ],
        [ 'spacing=i',  'wrapped line spacing',              { default => 1 }, ],
        [ 'colors=i',   'number of colors',                  { default => 15 }, ],
        [ 'section=i', 'start section', { default => 1, hidden => 1 }, ],
        [ 'outgroup',  'alignments have an outgroup', ],
        [ 'noindel',   'omit indels', ],
        [ 'nosingle',  'omit singleton SNPs and indels', ],
        [ 'nocomplex', 'omit complex SNPs and indels', ],
        { show_defaults => 1, }
    );
}

sub usage_desc {
    return "fasops xlsx [options] <infile>";
}

sub description {
    my $desc;
    $desc .= ucfirst(abstract) . ".\n";
    $desc .= <<'MARKDOWN';

* <infiles> are paths to axt files, .axt.gz is supported
* infile == stdin means reading from STDIN

MARKDOWN

    return $desc;
}

sub validate_args {
    my ( $self, $opt, $args ) = @_;

    if ( @{$args} != 1 ) {
        my $message = "This command need one input file.\n\tIt found";
        $message .= sprintf " [%s]", $_ for @{$args};
        $message .= ".\n";
        $self->usage_error($message);
    }
    for ( @{$args} ) {
        next if lc $_ eq "stdin";
        if ( !Path::Tiny::path($_)->is_file ) {
            $self->usage_error("The input file [$_] doesn't exist.");
        }
    }

    if ( !exists $opt->{outfile} ) {
        $opt->{outfile} = Path::Tiny::path( $args->[0] )->absolute . ".xlsx";
    }

    if ( $opt->{colors} ) {
        $opt->{colors} = List::Util::min( $opt->{colors}, 15 );
    }
}

sub execute {
    my ( $self, $opt, $args ) = @_;

    #@type IO::Handle
    my $in_fh;
    if ( lc $args->[0] eq "stdin" ) {
        $in_fh = *STDIN{IO};
    }
    else {
        $in_fh = IO::Zlib->new( $args->[0], "rb" );
    }

    # Create workbook and worksheet objects
    #@type Excel::Writer::XLSX
    my $workbook = Excel::Writer::XLSX->new( $opt->{outfile} );

    #@type Excel::Writer::XLSX::Worksheet
    my $worksheet = $workbook->add_worksheet;

    my $format_of       = create_formats($workbook);
    my $max_name_length = 1;

    my $content = '';    # content of one block
    while (1) {
        last if $in_fh->eof and $content eq '';
        my $line = '';
        if ( !$in_fh->eof ) {
            $line = $in_fh->getline;
        }
        next if substr( $line, 0, 1 ) eq "#";

        if ( ( $line eq '' or $line =~ /^\s+$/ ) and $content ne '' ) {
            my $info_of = App::Fasops::Common::parse_block( $content, 1 );
            $content = '';

            my @full_names;
            my $seq_refs = [];

            for my $key ( keys %{$info_of} ) {
                push @full_names, $key;
                push @{$seq_refs}, $info_of->{$key}{seq};
            }

            if ( $opt->{length} ) {
                next if length $info_of->{ $full_names[0] }{seq} < $opt->{length};
            }

            print "Section [$opt->{section}]\n";
            $max_name_length = List::Util::max( $max_name_length, map {length} @full_names );

            # including indels and snps
            my $vars = App::Fasops::Common::get_vars( $seq_refs, $opt );
            $opt->{section} = App::Fasops::Common::paint_vars( $worksheet, $format_of, $opt, $vars,
                \@full_names );

        }
        else {
            $content .= $line;
        }
    }

    $in_fh->close;

    # format column
    $worksheet->set_column( 0, 0, $max_name_length + 1 );
    $worksheet->set_column( 1, $opt->{wrap} + 3, 1.6 );

    return;
}

# Excel formats
sub create_formats {

    #@type Excel::Writer::XLSX
    my $workbook = shift;

    my $format_of = {};

    # species name
    $format_of->{name} = $workbook->add_format(
        font => 'Courier New',
        size => 10,
    );

    # variation position
    $format_of->{pos} = $workbook->add_format(
        font     => 'Courier New',
        size     => 8,
        align    => 'center',
        valign   => 'vcenter',
        rotation => 90,
    );

    $format_of->{snp}   = {};
    $format_of->{indel} = {};

    # background
    my $bg_of = {};

    # 15
    my @colors = (
        22,    # Gray-25%, silver
        43,    # Light Yellow       0b001
        42,    # Light Green        0b010
        27,    # Lite Turquoise
        44,    # Pale Blue          0b100
        46,    # Lavender
        47,    # Tan
        24,    # Periwinkle
        49,    # Aqua
        51,    # Gold
        45,    # Rose
        52,    # Light Orange
        26,    # Ivory
        29,    # Coral
        31,    # Ice Blue

        #        30,    # Ocean Blue
        #        41,    # Light Turquoise, again
        #        48,    # Light Blue
        #        50,    # Lime
        #        54,    # Blue-Gray
        #        62,    # Indigo
    );

    for my $i ( 0 .. $#colors ) {
        $bg_of->{$i}{bg_color} = $colors[$i];

    }
    $bg_of->{unknown}{bg_color} = 9;    # White

    # snp base
    my $snp_fg_of = {
        'A' => { color => 58, },        # Dark Green
        'C' => { color => 18, },        # Dark Blue
        'G' => { color => 28, },        # Dark Purple
        'T' => { color => 16, },        # Dark Red
        'N' => { color => 8, },         # Black
        '-' => { color => 8, },         # Black
    };

    for my $fg ( keys %{$snp_fg_of} ) {
        for my $bg ( keys %{$bg_of} ) {
            $format_of->{snp}{"$fg$bg"} = $workbook->add_format(
                font   => 'Courier New',
                size   => 10,
                align  => 'center',
                valign => 'vcenter',
                %{ $snp_fg_of->{$fg} },
                %{ $bg_of->{$bg} },
            );
        }
    }
    $format_of->{snp}{'-'} = $workbook->add_format(
        font   => 'Courier New',
        size   => 10,
        align  => 'center',
        valign => 'vcenter',
    );

    for my $bg ( keys %{$bg_of} ) {
        $format_of->{indel}->{$bg} = $workbook->add_format(
            font   => 'Courier New',
            size   => 10,
            bold   => 1,
            align  => 'center',
            valign => 'vcenter',
            %{ $bg_of->{$bg} },
        );
    }

    return $format_of;
}

1;
